src/shared/codelens/localLambdaRunner.ts

```diff
@@ -532,72 +537,58 @@ const getEnvironmentVariables = (config: HandlerConfig): SAMTemplateEnvironmentV
     }
 }
 
-export async function attachDebugger(params: {
-    channeLogger: ChannelLogger,
-    debugConfig: DebugConfiguration,
-    telemetryService: TelemetryService,
-    startInvokeMillis: number,
-    runtime: string,
-}): Promise<{ success: boolean }> {
-    const channelLogger = params.channeLogger
-    const logger = params.channeLogger.logger
+export interface AttachDebuggerContext {
+    debugConfig: DebugConfiguration
+    maxAttempts: number
+    channelLogger: Pick<ChannelLogger, 'info' | 'error' | 'logger'>
+    onStartDebugging?: typeof vscode.debug.startDebugging
+    onRecordAttachDebuggerMetric?(attachResult: boolean | undefined, attempts: number, attachResultDate: Date): void
+    onWillRetry(): Promise<void>
+}
+
+export async function attachDebugger(
+    {
+        onStartDebugging = vscode.debug.startDebugging,
+        ...params
+    }: AttachDebuggerContext
+): Promise<{ success: boolean }> {
+    const channelLogger = params.channelLogger
+    const logger = params.channelLogger.logger
     logger.debug(`localLambdaRunner.attachDebugger: startDebugging with debugConfig: ${JSON.stringify(
         params.debugConfig,
         undefined,
         2
     )}`)
 
-    let isDebuggerAttached: boolean | undefined = false
+    let isDebuggerAttached: boolean | undefined
     let numAttempts = 0
-    let retryDelay = 1000
-    let shouldRetry = false
-    const retryEnabled = false // Change this to enable retry
+
+    channelLogger.info(
+        'AWS.output.sam.local.attaching',
+        'Attaching debugger to SAM Application...',
+    )
+
     do {
-        channelLogger.info(
-            'AWS.output.sam.local.attaching',
-            'Attempt number {0} to attach debugger to SAM Application...',
-            String(numAttempts + 1)
-        )
-        isDebuggerAttached = await vscode.debug.startDebugging(undefined, params.debugConfig)
         numAttempts += 1
-        if (isDebuggerAttached === undefined) {
-            isDebuggerAttached = false
-            shouldRetry = numAttempts < MAX_DEBUGGER_ATTEMPTS
-        } else if (!isDebuggerAttached) {
-            retryDelay *= 2
-
-            shouldRetry = retryEnabled && (numAttempts < MAX_DEBUGGER_ATTEMPTS)
-            if (shouldRetry) {
-                const currTime = new Date()
-                recordDebugAttachResult({
-                    telemetryService: params.telemetryService,
-                    attachResult: isDebuggerAttached,
-                    attempts: numAttempts,
-                    duration: currTime.getTime() - params.startInvokeMillis,
-                    runtime: params.runtime,
-                })
+        isDebuggerAttached = await onStartDebugging(undefined, params.debugConfig)
+
+        if (params.onRecordAttachDebuggerMetric) {
+            params.onRecordAttachDebuggerMetric(isDebuggerAttached, numAttempts, new Date())
```

Comment:
Do we want to record a metric after every attempt or only after final success/failure?

Reply:
If we record only after resolving retries, we capture the relevant information, I'll change that.